README
PDBC

The PDBC module provides a higher-level interface to the JDBC module.
(The JDBC module provides an interface to Java JDBC using Inline::Java.)
Specifically it provides a more DBI-like interface for JDBC.
The goal is for PDBC to define the DBI<->DBD API.

You'll need to install and test JDBC before you can use this module.

INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

COPYRIGHT AND LICENCE

Copyright (C) 2006 Tim Bunce

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

Enjoy!


Changes
Revision history for JDBC

0.01    5th October 2005
        First version released on an unsuspecting world.



t/test_init.pl
# perform various pre-test initializations

# initialise variables describing the driver we'll test with
# and setup CLASSPATH (which needs to be done before 'use PDBC')

use Carp;
use File::Path;

use Cwd;
my $dir = cwd;
$dir =~ s:/t$::;

$::PDBC_DRIVER_CLASS = "org.apache.derby.jdbc.EmbeddedDriver";
$::PDBC_DRIVER_URL   = "jdbc:derby:derbydb;create=true";
$::PDBC_DRIVER_JAR   = "$dir/drivers/db-derby-10.1.1.0.jar";
@::PDBC_DRIVER_DBFILES = qw(derbydb derbydb.log);
die "Driver $::PDBC_DRIVER_JAR not found" unless -s $::PDBC_DRIVER_JAR;
rmtree \@::PDBC_DRIVER_DBFILES, 0;

$ENV{CLASSPATH} = join ":", $::PDBC_DRIVER_JAR, split /:/, $ENV{CLASSPATH}, -1;

# make Carp more verbose

$Carp::Verbose = 1;

# optionally enable Inline::Java verbosity

$ENV{PERL_INLINE_JAVA_DEBUG} = 0; # 0=off, 1,2=useful, 3=noisy

#

$SIG{__DIE__} = sub {
        die $@ unless ref $@;
        my $msg = $@;
        # $@->printStrackTrace; # No public method 'printStrackTrace' defined for class 'main::java::lang::NoClassDefFoundError' at test.pl line 20
        $msg = sprintf "%s (%s)", $@->toString, $msg;
        die $msg;
} if 0;

1;


lib/PDBC.pm
package PDBC;

use warnings;
use strict;

=head1 NAME

PDBC - Perl 5 interface to Java JDBC with a more DBI-like interface

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 SYNOPSIS

    use PDBC;

    PDBC->load_driver("org.apache.derby.jdbc.EmbeddedDriver");

    my $con = PDBC->getConnection($url, "test", "test");

    my $s = $con->createStatement();

    $s->executeUpdate("create table foo (foo int, bar varchar(200), primary key (foo))");
    $s->executeUpdate("insert into foo (foo, bar) values (42,'notthis')");
    $s->executeUpdate("insert into foo (foo, bar) values (43,'notthat')");

    my $rs = $s->executeQuery("select foo, bar from foo");
    while ($rs->next) {
	my $foo = $rs->getInt(1);
	my $bar = $rs->getString(2);
	print "row: foo=$foo, bar=$bar\n";
    }

=head1 DESCRIPTION

This PDBC module provides an interface to the Java C<java.sql.*> and
C<javax.sql.*> JDBC APIs. It's actually a layer over the L<JDBC> module that
adds DBI-like methods.

=cut

use warnings;

use base qw(JDBC);

require PDBC::db;
require PDBC::st;

use Memoize;

memoize('_inject_methods');

our $debug = $ENV{PERL_PDBC_DEBUG} || 0;

sub _inject_methods { # memoized
    my ($to_pkg, $from_pkg) = @_;
    no strict 'refs';
    die "$to_pkg is not a package" unless %{$to_pkg.'::'};
    die "$from_pkg is not a package" unless %{$from_pkg.'::'};
    my $to_pkg_sub = $to_pkg . '__PDBC';
    #warn "Installing methods from $from_pkg into $to_pkg_sub\n";
    while (my ($name, $ref) = each %{$from_pkg.'::'}) {
        # import gets added automagically if the pkg is 'use'd.
        # similarly for BEGIN if the pkg contains a use.
        my $new_code_ref = *{$from_pkg.'::'.$name}{CODE}
                or next; # not a sub
        #warn "Installing $name into $to_pkg_sub\n";
        *{$to_pkg_sub . '::' . $name} = $new_code_ref;
    }
    my $super_methods_ref = \@{$from_pkg.'::SUPERS'};
    unshift @{$to_pkg_sub . '::ISA'}, $to_pkg;
    for my $meth (@$super_methods_ref) {
        eval qq{
            package $to_pkg_sub;
            sub SUPER__$meth { return shift->SUPER::$meth(\@_); }
        };
        die $@ if $@;
    }
    return $to_pkg_sub;
}

sub _rebless_handle {
    my ($h, $from_pkg) = @_;
    my $to_pkg_sub = PDBC::_inject_methods(ref $h, $from_pkg);
    bless $h, $to_pkg_sub;
    return $h;
}

sub getConnection {
    my $self = shift;
    my $con = $self->SUPER::getConnection(@_);
    use Devel::Peek;
    #Dump $con;
    $con = PDBC::_rebless_handle($con, 'PDBC::db', [qw(createStatement prepareStatement)]);
    #Dump $con;
    return $con
}


sub import {
    my $pkg = shift;
    #my $callpkg = caller($Exporter::ExportLevel);

    # call standard import to handle anything else
    local $Exporter::ExportLevel = $Exporter::ExportLevel + 1;
    return $pkg->SUPER::import(@_);
}

=head1 METHODS

=head2 load_driver

The load_driver() method is used to load a driver class.

  JDBC->load_driver($driver_class)

is equivalent to the Java:

  java.lang.Class.forName(driver_class).newInstance();

=cut

#sub load_driver;


=head1 FUNCTIONS

=head2 cast

=head2 caught

The cast(), caught() functions of JDBC are also optionally exported by the PDBC module.

=cut

=head1 IMPORTING CONSTANTS

Java JDBC makes use of constants defined in 

  import java.sql.*;
  ...
  stmt = con.prepareStatement(PreparedStatement.SELECT);

the package can also be specified with the C<import> which then avoids the need
to prefix the constant with the class:

  import java.sql.PreparedStatement;
  ...
  stmt = con.prepareStatement(SELECT);

In Perl the corresponding code can be either:

  use JDBC;
  ...
  $stmt = $con->prepareStatement($java::sql::PrepareStatement::SELECT);

or, the rather more friendly:

  use JDBC qw(:PreparedStatement);
  ...
  $stmt = $con->prepareStatement(SELECT);

When importing a JDBC class in this way the JDBC module only imports defined
scalars with all-uppercase names, and it turns them into perl constants so the
C<$> is no longer needed.

All constants in all the java.sql and javax.sql classes can be imported in this way.

=cut

1; # End of JDBC

__END__

=head1 SEE ALSO

L<JDBC>

=head1 AUTHOR

Tim Bunce, C<< <Tim.Bunce@pobox.com> >>

=head1 BUGS

Probably.

=head1 COPYRIGHT & LICENSE

Copyright 2006 Tim Bunce, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
# vim: sw=4


lib/PDBC/db.pm
package PDBC::db;

use NEXT;

# $h->DBI::db::begin_work([ \%attr ])  
# $h->DBI::db::clone([\%attr])  
# $h->DBI::db::column_info($catalog, $schema, $table, $column [, \%attr ])   IMA_NOT_FOUND_OKAY 
# $h->DBI::db::commit($;$)  IMA_END_WORK  IMA_NOT_FOUND_OKAY
# $h->DBI::db::connected()  IMA_STUB
# $h->DBI::db::data_sources([\%attr])  
# $h->DBI::db::disconnect($;$)   IMA_NOT_FOUND_OKAY
# $h->DBI::db::do($statement [, \%attr [, @bind_params ] ])   IMA_EXECUTE 
# $h->DBI::db::foreign_key_info($pk_catalog, $pk_schema, $pk_table, $fk_catalog, $fk_schema, $fk_table [, \%attr ])   IMA_NOT_FOUND_OKAY 
# $h->DBI::db::get_info($info_type)   IMA_NOT_FOUND_OKAY 
# $h->DBI::db::last_insert_id($catalog, $schema, $table_name, $field_name [, \%attr ])  IMA_NOT_FOUND_OKAY 
# $h->DBI::db::ping($;$)  IMA_KEEP_ERR 
# $h->DBI::db::prepare($statement [, \%attr])   
# $h->DBI::db::prepare_cached($statement [, \%attr [, $if_active ] ])   
# $h->DBI::db::preparse()  
# $h->DBI::db::primary_key($catalog, $schema, $table [, \%attr ])   
# $h->DBI::db::primary_key_info($catalog, $schema, $table [, \%attr ])   IMA_NOT_FOUND_OKAY 
# $h->DBI::db::quote($string [, $data_type ])  IMA_NO_TAINT_IN IMA_NO_TAINT_OUT 
# $h->DBI::db::quote_identifier($name [, ...] [, \%attr ])  IMA_NO_TAINT_IN IMA_NO_TAINT_OUT 
# $h->DBI::db::rollback($;$)  IMA_END_WORK  IMA_NOT_FOUND_OKAY
# $h->DBI::db::rows()  IMA_KEEP_ERR
# $h->DBI::db::selectall_arrayref($statement [, \%attr [, @bind_params ] ])  
# $h->DBI::db::selectall_hashref($statement, $keyfield [, \%attr [, @bind_params ] ])  
# $h->DBI::db::selectcol_arrayref($statement [, \%attr [, @bind_params ] ])  
# $h->DBI::db::selectrow_array($statement [, \%attr [, @bind_params ] ])  
# $h->DBI::db::selectrow_arrayref($statement [, \%attr [, @bind_params ] ])  
# $h->DBI::db::selectrow_hashref($statement [, \%attr [, @bind_params ] ])  
# $h->DBI::db::table_info($catalog, $schema, $table, $type [, \%attr ])   IMA_NOT_FOUND_OKAY 
# $h->DBI::db::tables($catalog, $schema, $table, $type [, \%attr ])   
# $h->DBI::db::take_imp_data($;$)  
# $h->DBI::db::type_info($data_type)   
# $h->DBI::db::type_info_all($;$)   IMA_NOT_FOUND_OKAY 

use PDBC::xx;

our @SUPERS = qw(createStatement prepareStatement);

sub _rebless_st_handle {
    return PDBC::_rebless_handle(shift, 'PDBC::st');
}

sub createStatement {
    my $self = shift;
    my $stmt = $self->SUPER__createStatement(@_);
    _rebless_st_handle($stmt);
    return $stmt;
}

sub prepareStatement {
    my $self = shift;
    my $stmt = $self->SUPER__prepareStatement(@_);
    _rebless_st_handle($stmt);
    return $stmt;
}


sub foo {
    warn "foo";
}

1;


lib/PDBC/rs.pm
package PDBC::rs;

#= $h->DBI::st::_get_fbav()  
#  $h->DBI::st::_set_fbav()  
#  $h->DBI::st::bind_col($column, \$var [, \%attr])  
#  $h->DBI::st::bind_columns(\$var1 [, \$var2, ...])  
#  $h->DBI::st::bind_param($parameter, $var [, \%attr])  
#  $h->DBI::st::bind_param_array($parameter, $var [, \%attr])  
#  $h->DBI::st::bind_param_inout($parameter, \$var, $maxlen, [, \%attr])  
#  $h->DBI::st::bind_param_inout_array($parameter, \@var, $maxlen, [, \%attr])  
#  $h->DBI::st::blob_copy_to_file($field, $filename_or_handleref)  
#  $h->DBI::st::blob_read($field, $offset, $len [, \$buf [, $bufoffset]])  
#  $h->DBI::st::cancel($;$)  IMA_NOT_FOUND_OKAY
#  $h->DBI::st::dump_results($maxfieldlen, $linesep, $fieldsep, $filehandle)  
#  $h->DBI::st::execute([@args])  IMA_COPY_STMT IMA_EXECUTE
#  $h->DBI::st::execute_array(\%attribs [, @args])  IMA_COPY_STMT IMA_EXECUTE
#  $h->DBI::st::execute_for_fetch($fetch_sub [, $tuple_status])  IMA_COPY_STMT IMA_EXECUTE
#- $h->DBI::st::fetch()  
#  $h->DBI::st::fetchall_arrayref([ $slice [, $max_rows]])  
#  $h->DBI::st::fetchall_hashref($key_field)  
#- $h->DBI::st::fetchrow()  
#= $h->DBI::st::fetchrow_array()  
#= $h->DBI::st::fetchrow_arrayref()  
#  $h->DBI::st::fetchrow_hashref()  
#  $h->DBI::st::finish($;$)  
#  $h->DBI::st::more_results($;$)  
#  $h->DBI::st::rows()  IMA_KEEP_ERR

sub _get_fbav {
    # XXX should belong to DBI not DBD since the DBI should handle
    # column binding

    my $rs = shift;
    # XXX should return the same row ref each time
    # (to which bound columns would be aliases)
    # for now we'll just return a new one each time
    my $cols = $rs->getMetaData()->getColumnCount();
    my @row;
    $#row = $cols-1; # [0..$cols-1]
    return \@row;
}


sub fetchrow_arrayref {
    my $rs = shift;
    $rs->next or return;
    my $row = $rs->_get_fbav();
    for (0..@$row-1) {
        # assign to each individually so column bindings work
        $row->[$_] = $rs->getString($_+1)
    }
    return \@row;
}


sub fetchrow_array {
    my $h = shift;
    my $row = $h->fetchrow_arrayref or return;
    return @$row;
}


return 1;


lib/PDBC/xx.pm
package PDBC::xx;

# package for methods common to drh, dbh & sth

use base qw(Exporter);

# export everything
our @EXPORT = qw();

# $h->DBI::common::debug([$debug_level])  IMA_KEEP_ERR
# $h->DBI::common::dump_handle([$message [, $level]])  IMA_KEEP_ERR
# $h->DBI::common::err()  IMA_KEEP_ERR
# $h->DBI::common::errstr()  IMA_KEEP_ERR
# $h->DBI::common::func()  IMA_FUNC_REDIRECT IMA_KEEP_ERR
# $h->DBI::common::parse_trace_flag($name)  IMA_KEEP_ERR 
# $h->DBI::common::parse_trace_flags($flags)  IMA_KEEP_ERR 
# $h->DBI::common::private_data($;$)  IMA_KEEP_ERR
# $h->DBI::common::set_err($err, $errmsg [, $state, $method, $rv])  IMA_NO_TAINT_IN
# $h->DBI::common::state()  IMA_KEEP_ERR
# $h->DBI::common::swap_inner_handle($h [, $allow_reparent ])  
# $h->DBI::common::trace([$trace_level, [$filename]])  IMA_KEEP_ERR
# $h->DBI::common::trace_msg($message_text [, $min_level ])  IMA_KEEP_ERR

1;


lib/JDBC.pm
package JDBC;

use warnings;
use strict;

=head1 NAME

JDBC - Perl 5 interface to Java JDBC (via Inline::Java)

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 SYNOPSIS

    use JDBC;

    JDBC->load_driver("org.apache.derby.jdbc.EmbeddedDriver");

    my $con = JDBC->getConnection($url, "test", "test");

    my $s = $con->createStatement();

    $s->executeUpdate("create table foo (foo int, bar varchar(200), primary key (foo))");
    $s->executeUpdate("insert into foo (foo, bar) values (42,'notthis')");
    $s->executeUpdate("insert into foo (foo, bar) values (43,'notthat')");

    my $rs = $s->executeQuery("select foo, bar from foo");
    while ($rs->next) {
	my $foo = $rs->getInt(1);
	my $bar = $rs->getString(2);
	print "row: foo=$foo, bar=$bar\n";
    }

=head1 DESCRIPTION

This JDBC module provides an interface to the Java JDBC APIs.

=cut

our @ISA = qw(Exporter java::sql::DriverManager);

{   # the Inline package needs to be use'd in main in order to
    # get the studied classes to be rooted in main
    package main;
    use Inline ( Java => q{ }, AUTOSTUDY => 1 );
}

use Inline::Java qw(cast caught study_classes);

our @EXPORT_OK = qw(cast caught study_classes);

our $debug = $ENV{PERL_JDBC_DEBUG} || 0;

# Could use package aliasing (*{$alias . "::"} = \*{$orig . "::"}) to shortcut these
# so $ResultSet::FOO would work, but we'd have to pollute ResultSet:: globally.
# Probably best to "use JDBC qw(:ResultSet);" and have that import all the
# java.sql.ResultSet.* constants (defined scalars with all-caps names)

#java.sql.ParameterMetaData 
my @classes = (qw(
java.sql.Array 
java.sql.BatchUpdateException 
java.sql.Blob 
java.sql.CallableStatement 
java.sql.Clob 
java.sql.Connection 
java.sql.DataTruncation 
java.sql.DatabaseMetaData 
java.sql.Date 
java.sql.Driver 
java.sql.DriverManager 
java.sql.DriverPropertyInfo 
java.sql.PreparedStatement 
java.sql.Ref 
java.sql.ResultSet
java.sql.ResultSetMetaData 
java.sql.SQLData 
java.sql.SQLException 
java.sql.SQLInput 
java.sql.SQLOutput 
java.sql.SQLPermission 
java.sql.SQLWarning 
java.sql.Savepoint 
java.sql.Statement 
java.sql.Struct 
java.sql.Time 
java.sql.Timestamp 
java.sql.Types 
javax.sql.ConnectionEvent 
javax.sql.ConnectionEventListener 
javax.sql.ConnectionPoolDataSource 
javax.sql.DataSource 
javax.sql.PooledConnection 
javax.sql.RowSet 
javax.sql.RowSetEvent 
javax.sql.RowSetInternal 
javax.sql.RowSetListener 
javax.sql.RowSetMetaData 
javax.sql.RowSetReader 
javax.sql.RowSetWriter 
javax.sql.XAConnection 
javax.sql.XADataSource 

));

warn "studying @classes\n" if $debug;
study_classes(\@classes, 'main');

warn "running\n" if $debug;

=head1 METHODS

=head2 load_driver

The load_driver() method is used to load a driver class.

  JDBC->load_driver($driver_class)

is equivalent to the Java:

  java.lang.Class.forName(driver_class).newInstance();

=cut

sub load_driver {
    my ($self, $class) = @_;
    study_classes([$class], 'main');
}

# override getDrivers to return an Enumeration (not private class)

sub getDrivers {
    return cast('java.util.Enumeration', shift->SUPER::getDrivers)
}

=head1 FUNCTIONS

=head2 cast

=head2 caught

=head2 study_classes

The cast(), caught(), and study_classes() functions of Inline::Java are also
optionally exported by the JDBC module.

=cut

*java_cast = \&cast;


1; # End of JDBC

__END__

=head1 SEE ALSO
=head1 SEE ALSO

L<Inline::Java>

=head1 AUTHOR

Tim Bunce, C<< <Tim.Bunce@pobox.com> >>

=head1 BUGS

Firstly try to determine if the problem is related to the JDBC module itself
or, more likely, the underlying Inline::Java module.

Please report any bugs or feature requests for JDBC to
C<bug-jdbc@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=JDBC>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

Please report any bugs or feature requests for Inline::Java to the Inline::Java
mailing list.

C<Inline::Java>'s mailing list is <inline@perl.org>.
To subscribe, send an email to <inline-subscribe@perl.org>

C<Inline::Java>'s home page is http://inline.perl.org/java/

=head1 ACKNOWLEDGEMENTS

Thanks to Patrick LeBoutillier for creating Inline::Java.

=head1 COPYRIGHT & LICENSE

Copyright 2005 Tim Bunce, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
# vim: sw=4
